Release/175.0.0 (#4525)
MetaMask/core: packages/assets-controllers/CHANGELOG.md

## [Unreleased]

> ## [35.1.0]

Reviewer: Because of the breaking change below, should this be 36.0.0?

### Added

- Add optional `topBid` property to the `NftMetadata` type. This property must be of type `TopBid`. ([#4522](https://github.com/MetaMask/core/pull/4522))
- Add optional `floorAsk` property to the `TokenCollection` type. This property must be of type `FloorAskCollection`. ([#4522](https://github.com/MetaMask/core/pull/4522))
- Add linea mainnet support to nft detection supported networks ([#4515](https://github.com/MetaMask/core/pull/4515))

> - Fetch NFT collections data from the NFT-API `getCollections` API when calling the `detectNfts` method of `NftDetectionController`, and the `updateNftMetadata` and `watchNft` methods of `NftController`. ([#4443](https://github.com/MetaMask/core/pull/4443))
> 
> ### Changed
> 

Reviewer 2: This looks like it might be a change in behavior for the affected class methods, not an example of the package API's surface area being extended (see https://github.com/MetaMask/core/blob/main/docs/reviewing-release-prs.md#additions).

Would you agree that we are returning more data to the consumer (or to other intermediate call sites), but we're not expanding the interface with which the consumer is expected to interact when calling these methods?

```suggestion

### Changed

- Fetch NFT collections data from the NFT-API `getCollections` API when calling the `detectNfts` method of `NftDetectionController`, and the `updateNftMetadata` and `watchNft` methods of `NftController`. ([#4443](https://github.com/MetaMask/core/pull/4443))
```

Author: Agree, when clients want to interact with the new responses they will interact with 
`import { Nft } from '@metamask/assets-controllers';` where all the newly fetched data is optional and new access to the new data will treat new values as optional;

Reviewer 2: If so, shouldn't this entry be recategorized as a `Changed` entry?

Author: Hmmm, I dnt have a strong opinion on this, I thought we are in the case of adding optional properties to an existing type in the Additions right? [section](https://github.com/MetaMask/core/blob/main/docs/reviewing-release-prs.md#additions)

Where clients in future would also treat this new data as optional; so we are not creating any new dependencies.
Just adding few more optional values that if exists should be displayed, if not, they will be ignored.

I can update it to `Changed` if you think that makes more sense

Reviewer 2: The type changes and the method behavior changes seem like separate updates. How would you feel about splitting this up into an Added and a Changed entry?

```suggestion
- The `Collection` type is expanded to include the following 'string'-type optional properties: `contractDeployedAt`, `creator`, `ownerCount`, and an optional property `topBid` of the type `TopBid & { sourceDomain?: string; }`. ([#4443](https://github.com/MetaMask/core/pull/4443))

### Changed

- Fetch NFT collections data from the OpenSea API `Get Collections` endpoint when calling the `detectNfts` method of `NftDetectionController`, and the `updateNftMetadata` and `watchNft` methods of `NftController`. ([#4443](https://github.com/MetaMask/core/pull/4443))
```

Author: Updated 🙏

Author: I only put NFT-API instead of OpenSea API because it is more accurate;

- Change attributes type in `NftMetadata` to `Attributes[]` ([#4522](https://github.com/MetaMask/core/pull/4522))
- Bump `@metamask/utils` to `^9.0.0` ([#4516](https://github.com/MetaMask/core/pull/4516))
- Bump `@metamask/rpc-errors` to `^6.3.1` ([#4516](https://github.com/MetaMask/core/pull/4516))

> 

Reviewer: In the Changelog section of the PR description, the `attributes` type change is prepended with "FIXED". Should this go under the `Fixed` header?

In addition, since `@metamask/assets-controllers` exports this type, should we mark this as a breaking change? 

``` suggestion
- Bump `@metamask/utils` to `^9.0.0` ([#4516](https://github.com/MetaMask/core/pull/4516))
- Bump `@metamask/rpc-errors` to `^6.3.1` ([#4516](https://github.com/MetaMask/core/pull/4516))

### Fixed

- **BREAKING:** Change `attributes` type in `NftMetadata` to `Attributes[]` ([#4522](https://github.com/MetaMask/core/pull/4522))
```

Author: Thanks @mcmire for the review 🙏 
I checked how clients are using this attribute,
I see extension is not using it so far, i have an ongoing PR that incorporates the Attributes[];
Same goes for mobile;

So i thought maybe it is not necessary to mark it as breaking;

I could update it to Breaking if you think it is better!

Reviewer: @sahar-fehri It's good to check whether the extension or mobile app is currently using this, so thanks for doing that. I think it would still be good practice, however, to mark this as breaking, so that if the clients encounter an unexpected error while upgrading in the near future or if we simply want to look back at history in the far future, we are doing our part to share that information now and aren't hiding it.

Author: That makes sense! i updated the version to `36.0.0`

Reviewer: Thanks!

## [Unreleased]

## [35.1.0]

### Added

- Added optional `topBid` type in `NftMetadata` ([#4522](https://github.com/MetaMask/core/pull/4522))

> - Added optional `floorAsk` in `TokenCollection` type ([#4522](https://github.com/MetaMask/core/pull/4522))

Author: Resolving this now after our discussion ^^

- Added linea mainnet support to nft detection supported networks ([#4515](https://github.com/MetaMask/core/pull/4515))

> - Added a batch call to get collections API in detectNfts function ([#4443](https://github.com/MetaMask/core/pull/4443))
> - Added a call to get collections API inside getNftInformationFromApi ([#4443](https://github.com/MetaMask/core/pull/4443))
> 
> ### Changed
> 

Reviewer 2: ```suggestion

### Changed

- Fetch NFT collections data from the NFT-API `getCollections` API when calling the `detectNfts` method of `NftDetectionController`, and the `updateNftMetadata` and `watchNft` methods of `NftController`. ([#4443](https://github.com/MetaMask/core/pull/4443))
```

- Recategorized to `Changed`
- Replaced reference to private class method with consumer-facing changes
- Is there any additional information we can provide about this API?

Reviewer 2: If the source of the `getCollections` API is expected to be very clear to the consumers of this package, then we shouldn't need to add more information. Looks like this is the case here since it appears the NFT-API being referenced was already in use.

Author: Apologies @MajorLift i'm not sure i see what you meant here ☝️  could you elaborate? 🙏 

If by "NFT-API being referenced was already in use" you were referring to [this](https://github.com/MetaMask/core/blob/562e6a1aa44f1aa66b794d41590afc53e14dbddc/packages/assets-controllers/src/NftDetectionController.ts#L546) API, then yes it was already being used. 
With the new NFT UI designs we wanted to fetch more data that were available on the [getCollections](https://github.com/MetaMask/core/blob/562e6a1aa44f1aa66b794d41590afc53e14dbddc/packages/assets-controllers/src/NftDetectionController.ts#L665) API and then display that data if it exists in the UI.

So we added the call to [getCollections](https://github.com/MetaMask/core/blob/562e6a1aa44f1aa66b794d41590afc53e14dbddc/packages/assets-controllers/src/NftDetectionController.ts#L665) api after calling [getUserNFts](https://github.com/MetaMask/core/blob/562e6a1aa44f1aa66b794d41590afc53e14dbddc/packages/assets-controllers/src/NftDetectionController.ts#L629) api to fill in any missing fields;

Reviewer 2: My concern is that the consumer might not know where this new `getCollections` API is coming from without digging through the code. It's unclear whether we're pulling from opensea or some internal endpoint we have already been using, or if we're pulling from a completely new external service. "NFT-API" doesn't seem to offer much clarification, as there are quite a few services with that name. 

Just saying it would be nice to have something like "OpenSea `getCollections` API" explicitly stated in the entry, unless just "`getCollections` API" is clear enough for our downstream consumers (who may or may not work for MetaMask).

Reviewer 2: Superseded by https://github.com/MetaMask/core/pull/4525#discussion_r1678246446

> - Added optional `topBid` type in `NftMetadata` ([#4522](https://github.com/MetaMask/core/pull/4522))
> - Added optional `floorAsk` in `TokenCollection` type ([#4522](https://github.com/MetaMask/core/pull/4522))
> - Added linea mainnet support to nft detection supported networks ([#4515](https://github.com/MetaMask/core/pull/4515))
> - Added a batch call to get collections API in detectNfts function ([#4443](https://github.com/MetaMask/core/pull/4443))
> - Added a call to get collections API inside getNftInformationFromApi ([#4443](https://github.com/MetaMask/core/pull/4443))
> 
> ### Changed
> 
> - Changed attributes type in `NftMetadata` to `Attributes[]` ([#4522](https://github.com/MetaMask/core/pull/4522))

Reviewer 2: nit: Changelog entries should start with a present-tense verb (https://common-changelog.org/#241-change)

```suggestion
- Add optional `topBid` type in `NftMetadata` ([#4522](https://github.com/MetaMask/core/pull/4522))
- Add optional `floorAsk` in `TokenCollection` type ([#4522](https://github.com/MetaMask/core/pull/4522))
- Add linea mainnet support to nft detection supported networks ([#4515](https://github.com/MetaMask/core/pull/4515))
- Add a batch call to get collections API in detectNfts function ([#4443](https://github.com/MetaMask/core/pull/4443))
- Add a call to get collections API inside getNftInformationFromApi ([#4443](https://github.com/MetaMask/core/pull/4443))

### Changed

- Change attributes type in `NftMetadata` to `Attributes[]` ([#4522](https://github.com/MetaMask/core/pull/4522))
```

> - Bump `@metamask/utils` to `^9.0.0` ([#4508](https://github.com/MetaMask/core/pull/4516))
> - Bump `@metamask/rpc-errors` to `^6.3.1` ([#4498](https://github.com/MetaMask/core/pull/4516))

Reviewer 2: ```suggestion
- Bump `@metamask/utils` to `^9.0.0` ([#4516](https://github.com/MetaMask/core/pull/4516))
- Bump `@metamask/rpc-errors` to `^6.3.1` ([#4516](https://github.com/MetaMask/core/pull/4516))
```

- Add optional `floorAsk` property to the `TokenCollection` type. This property must be of type `FloorAskCollection`. ([#4522](https://github.com/MetaMask/core/pull/4522))
- Add linea mainnet support to nft detection supported networks ([#4515](https://github.com/MetaMask/core/pull/4515))
- The `Collection` type is expanded to include the following 'string'-type optional properties: `contractDeployedAt`, `creator`, `ownerCount`, and an optional property `topBid` of the type `TopBid & { sourceDomain?: string; }`. ([#4443](https://github.com/MetaMask/core/pull/4443))

### Changed

- Fetch NFT collections data from the NFT-API `Get Collections` endpoint when calling the `detectNfts` method of `NftDetectionController`, and the `updateNftMetadata` and `watchNft` methods of `NftController`. ([#4443](https://github.com/MetaMask/core/pull/4443))
- Bump `@metamask/utils` to `^9.0.0` ([#4516](https://github.com/MetaMask/core/pull/4516))
- Bump `@metamask/rpc-errors` to `^6.3.1` ([#4516](https://github.com/MetaMask/core/pull/4516))

### Fixed

> - **BREAKING:** Change `attributes` type in `NftMetadata` to `Attributes[]` ([#4522](https://github.com/MetaMask/core/pull/4522))

Reviewer 2: ```suggestion
- **BREAKING:** The `attributes` property of the `NftMetadata` type must be of type `Attributes[]` ([#4522](https://github.com/MetaMask/core/pull/4522))
  - This property has been typed as `Attributes` since `@metamask/assets-controllers` `v28.0.0`.
```

Reviewer 2: Was `Attributes` always the wrong type to have here, or are we changing this due to a newly-introduced upstream change?

Reviewer 2: That's good to know, but looks like they're both still in the affected range (v28-v36) since extension is at v34, and mobile is at v30. If they, or any other future consumer in the affected range decide to use `attributes`, they'll need to patch its type to `Attributes[]`. This entry should make them aware of this fact, and whether they should apply this patch, or simply update to `>=36.0.0`.

Author: Totally, makes sense 🙏

- Use Ethers for AssetsContractController ([#845](https://github.com/MetaMask/core/pull/845))

> 
> [Unreleased]: https://github.com/MetaMask/core/compare/@metamask/assets-controllers@35.1.0...HEAD
> [35.1.0]: https://github.com/MetaMask/core/compare/@metamask/assets-controllers@35.0.0...@metamask/assets-controllers@35.1.0

Reviewer 3: As part of Elliot's suggestion to make a major release this change should be made too
```suggestion

[Unreleased]: https://github.com/MetaMask/core/compare/@metamask/assets-controllers@36.0.0...HEAD
[36.0.0]: https://github.com/MetaMask/core/compare/@metamask/assets-controllers@35.0.0...@metamask/assets-controllers@36.0.0
```

Author: Thank you @cryptodev-2s  for pointing this out! 🙏
